internal/engine/ddl/foreign_procedure.go

```diff
@@ -0,0 +1,340 @@
+    IF _expected_input_types[1] != 'TEXT' THEN
+        RAISE EXCEPTION 'Procedure "%" expects input type "TEXT", but got "%"', _procedure, _expected_input_types[1];
+    END IF;
+
+    IF _expected_input_types[2] != 'INT' THEN
+        RAISE EXCEPTION 'Procedure "%" expects input type "INT", but got "%"', _procedure, _expected_input_types[2];
+    END IF;
+
+    -- 5. our return types match the expected return types
+    IF array_length(_expected_return_types, 1) != 1 THEN
+        RAISE EXCEPTION 'Procedure "%" expects exactly one return type, but got %', _procedure, array_length(_expected_return_types, 1);
+    END IF;
+
+    -- since _out_1 is text, we want to ensure that the return type is text
+    IF _expected_return_types[1] != 'TEXT' THEN
+        RAISE EXCEPTION 'Procedure "%" expects return type "TEXT", but got "%"', _procedure, _expected_return_types[1];
+    END IF;
+
+    -- we now call the procedure. we prefix ds_ to the dbid, as per the Kwil rules.
+    EXECUTE format('SELECT * FROM ds_%I.%I($1, $2)', _dbid, _procedure) INTO _out_1 USING _arg1, _arg2;
+
+	-- or, to return a table, we can do:
+	-- RETURN QUERY EXECUTE format('SELECT * FROM ds_%I.%I($1, $2)', _dbid, _procedure) USING _arg1, _arg2;
+
+END;
+$$ LANGUAGE plpgsql;
+*/
+
+// This is implicitly
+// coupled to the schema defined in internal/engine.execution/queries.go, and therefore is implicitly
+// a circular dependency. I am unsure how to resolve this, but am punting on it for now since the structure
+// of the new parts of the engine are still in flux.
+
+// GenerateForeignProcedure generates a plpgsql function that allows the schema to dynamically
+// call procedures in other schemas, expecting certain inputs and return values. It will prefix
+// the generated function with _fp_ (for "foreign procedure").
+func GenerateForeignProcedure(proc *types.ForeignProcedure) (string, error) {
+	str := strings.Builder{}
+
+	// first write the header
+	str.WriteString(fmt.Sprintf(`CREATE OR REPLACE FUNCTION _fp_%s(_dbid TEXT, _procedure TEXT`, proc.Name))
+
+	// we now need to format the inputs. Inputs will be named _arg1, _arg2, etc.
+	// we start at 1 since postgres is 1-indexed.
+	argList := make([]string, len(proc.Parameters))
+	for i, in := range proc.Parameters {
+		pgStr, err := in.PGString()
+		if err != nil {
+			return "", err
+		}
+		name := fmt.Sprintf("_arg%d", i+1)
+		str.WriteString(fmt.Sprintf(", %s %s", name, pgStr))
+		argList[i] = name
+	}
+
+	var outList []string
+	// if there are non-table outputs, we need to format them.
+	// we can ignore the name of the output, since it is not a table
+	if proc.Returns != nil && !proc.Returns.IsTable {
+		for i, out := range proc.Returns.Fields {
+			str.WriteString(", OUT ")
+			pgStr, err := out.Type.PGString()
+			if err != nil {
+				return "", err
+			}
+
+			str.WriteString("_out_")
+			name := fmt.Sprintf("_out_%d", i+1)
+			str.WriteString(name)
```

Comment:
Maybe I'm reading this wrong because it's Friday afternoon, but this looks like it's going to write `_out__out_X`.

Reply:
Lol yes this is a bug